### ISSUE TITLE
[lld][ELF][RISCV] medlow absolute HI20 overflow is reported before GP relaxation can make the sequence valid

### ISSUE THREAD

**Issue by XiaobingHou1219** [llvm/llvm-project#201036]
I reduced this to RISC-V ELF linker inputs that do not depend on a compiler frontend. The testcase is small enough to inspect directly, and the GNU/binutils control files included with the reproducer describe the relocation or layout shape that lld is expected to preserve.

The behavior I see is not just a diagnostic wording difference. The final link either succeeds with the wrong relocation/layout metadata, rejects a link that the corresponding GNU control treats as valid, or crashes during linker bookkeeping.

### `lld.riscv.medlow_abs_hi20_overflow_not_prevented_by_gp_relaxation`
- Case id: `20260527-lld-medlow-overflow-before-gp-relax`
- Route: `confirmed_locally_from_binutils_code_model_relax_medlow_replay`
- Arch/ABI: `rv64i` / `lp64`
- Intent: Link a medlow RV64 executable whose absolute `%hi/%lo` data references become legal only after gp relaxation.
- What happens: GNU ld 2.45 links the legal input and rewrites both absolute data accesses to `addi t0,gp,-0x10` and `addi t0,gp,-0x8`, but ld.lld 21.1.3 and 22.1.0 reject the same link with `R_RISCV_HI20 out of range` before applying gp relaxation.
- Evidence in this bundle: 3 clean reproductions under hunt\\verify\\lld_medlow_abs_hi20_overflow_not_prevented_by_gp_relaxation\\run1..run3 show stable assembly success for medlow.o, stable GNU ld outputs whose bfd.objdump.txt begins with `addi t0,gp,-0x10` and `addi t0,gp,-0x8`, and stable ld.lld 21.1.3/22.1.0 failures whose stderr files report out-of-range `R_RISCV_HI20` against `symbolL` and `symbolG`.

### `lld.riscv.medlow_abs_hi20_overflow_not_prevented_by_gp_relaxation`
Verification directories in the package:

- `run1/`: 14 files; notable: bfd.objdump.err, bfd.objdump.txt, bfd.rc.txt, bfd.stderr.txt, bfd.stdout.txt, case.s, layout.ld, lld21.rc.txt, lld21.stderr.txt, lld21.stdout.txt, lld22.rc.txt, lld22.stderr.txt
- `run2/`: 14 files; notable: bfd.objdump.err, bfd.objdump.txt, bfd.rc.txt, bfd.stderr.txt, bfd.stdout.txt, case.s, layout.ld, lld21.rc.txt, lld21.stderr.txt, lld21.stdout.txt, lld22.rc.txt, lld22.stderr.txt
- `run3/`: 14 files; notable: bfd.objdump.err, bfd.objdump.txt, bfd.rc.txt, bfd.stderr.txt, bfd.stdout.txt, case.s, layout.ld, lld21.rc.txt, lld21.stderr.txt, lld21.stdout.txt, lld22.rc.txt, lld22.stderr.txt

<details>
<summary>Reduced source/control files</summary>

### `README.md`

```markdown
Intent:
- Link a medlow RV64 executable whose absolute `%hi/%lo` data references only become legal after gp relaxation.

Expected behavior:
- The linker should relax the two absolute accesses to `gp`-relative `addi` forms and complete the link, as GNU `ld` does.

Observed defect:
- `ld.lld 21.1.3/22.1.0` reject the legal input with `R_RISCV_HI20 out of range` before applying gp relaxation, while GNU `ld 2.45` rewrites both accesses to `addi t0,gp,...`.

Verification:
- `hunt\verify\lld_medlow_abs_hi20_overflow_not_prevented_by_gp_relaxation\run1..run3`
```

### `case.s`

```asm
	.text
	.global _start
_start:

.ifdef __medany__
	.option pic
.ifdef __undefweak__
	# Refer to undefined weak symbol by GOT_PCREL.
	la	t0, symbolW
	.option nopic
.else
	# Refer to global data symbol by GOT_PCREL.
	la	t0, symbolG
	.option nopic
	# Refer to local data symbol by PCREL.
	lla	t0, symbolL
	# Refer to non-pic data global symbol by PCREL.
	la	t0, symbolG
.endif
.endif

.ifdef __medlow__
.ifdef __undefweak__
	# Refer to undefined weak symbol by absolutely access.
	lui	t0, %hi(symbolW)
	addi	t0, t0, %lo(symbolW)
.else
	# Refer to local data symbol by absolutely access.
	lui	t0, %hi(symbolL)
	addi	t0, t0, %lo(symbolL)
	# Refer to global data symbol by absolutely access.
	lui	t0, %hi(symbolG)
	addi	t0, t0, %lo(symbolG)
.endif
.endif
	.size   _start, .-_start

	.data
	.global symbolG
symbolL:
	.dword	0x1111222233334444
symbolG:
	.dword	0x5555666677778888

.ifdef __undefweak__
	.weak	symbolW
.endif
```

### `layout.ld`

```ld
/* text: in the 32-bit address range.
   data: far away from text. */

ENTRY(_start)
MEMORY
{
	rom (rx)  : ORIGIN = 0x10000, LENGTH = 0x1000
	ram (!rx) : ORIGIN = 0x800000000, LENGTH = 0x1000
}
SECTIONS {
	.text : {
		*(.text*)
	} >rom

	.data : {
		*(.data*)
	} >ram

	__global_pointer$ = .;

	.got : {
		*(.got*)
	} > ram
}
```

</details>

**Comment by lenary** [comment]
The examples are not actually following the medlow code model, which only allows symbol references in the low 2GB and high 2GB of the address range (as can be materialised using an absolute signed 32-bit address): https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#medium-low-code-model 

The data section here is not within that range, so the error from LLD is entirely justified, regardless of bfd.ld's behaviour.

**Comment by dwblaikie** [comment]
Closing as not planned. Refer to issue #202459 for context.
